operator/pkg/gateway-api/status.go

```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package gateway_api
+
+import (
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func newCondition(conditionType string, status metav1.ConditionStatus, reason, msg string, lastTransitionTime time.Time, observedGeneration int64) metav1.Condition {
+	return metav1.Condition{
+		Type:               conditionType,
+		Status:             status,
+		Reason:             reason,
+		Message:            msg,
+		LastTransitionTime: metav1.NewTime(lastTransitionTime),
+		ObservedGeneration: observedGeneration,
+	}
+}
+
+// merge combines the provided conditions with the existing conditions.
+func merge(existingConditions []metav1.Condition, updates ...metav1.Condition) []metav1.Condition {
+	var additions []metav1.Condition
+	for i, update := range updates {
+		found := false
+		for j, cond := range existingConditions {
+			if cond.Type == update.Type {
+				found = true
+				if conditionChanged(cond, update) {
+					existingConditions[j].Status = update.Status
+					existingConditions[j].Reason = update.Reason
+					existingConditions[j].Message = update.Message
+					existingConditions[j].ObservedGeneration = update.ObservedGeneration
+					existingConditions[j].LastTransitionTime = update.LastTransitionTime
+					break
```

Comment:
Since we should have at most one condition for each type, I think we should move this `break` outside of the `if` scope. Doing so, we will immediately break out of the inner loop even when we'll found a condition that does not need any update.